{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os import path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "curr_dir_path = str(pathlib.Path().absolute())\n",
    "data_path = curr_dir_path + \"/Data/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  chr_name  source        type    start      end strand                  0  \\\n",
      "0    chr21  HAVANA        gene  5011799  5017145      +  ENSG00000279493.1   \n",
      "1    chr21  HAVANA  transcript  5011799  5017145      +  ENSG00000279493.1   \n",
      "2    chr21  HAVANA        exon  5011799  5011874      +  ENSG00000279493.1   \n",
      "3    chr21  HAVANA         CDS  5011799  5011874      +  ENSG00000279493.1   \n",
      "4    chr21  HAVANA        exon  5012548  5012687      +  ENSG00000279493.1   \n",
      "\n",
      "                     1                            2                        3  \\\n",
      "0         otein_coding       gene_name \"FP565260.4\"                  level 2   \n",
      "1   \"ENST00000624081.1   gene_type \"protein_coding\"   gene_name \"FP565260.4\"   \n",
      "2   \"ENST00000624081.1   gene_type \"protein_coding\"   gene_name \"FP565260.4\"   \n",
      "3   \"ENST00000624081.1   gene_type \"protein_coding\"   gene_name \"FP565260.4\"   \n",
      "4   \"ENST00000624081.1   gene_type \"protein_coding\"   gene_name \"FP565260.4\"   \n",
      "\n",
      "   ...                         12                                   13  \\\n",
      "0  ...                       None                                 None   \n",
      "1  ...   tag \"appris_principal_1\"   havana_gene \"OTTHUMG00000189354.1\"   \n",
      "2  ...        tag \"mRNA_start_NF\"                          tag \"basic\"   \n",
      "3  ...        tag \"mRNA_start_NF\"                          tag \"basic\"   \n",
      "4  ...        tag \"mRNA_start_NF\"                          tag \"basic\"   \n",
      "\n",
      "                                          14  \\\n",
      "0                                       None   \n",
      "1   havana_transcript \"OTTHUMT00000479422.1\"   \n",
      "2                   tag \"appris_principal_1\"   \n",
      "3                   tag \"appris_principal_1\"   \n",
      "4                   tag \"appris_principal_1\"   \n",
      "\n",
      "                                    15  \\\n",
      "0                                 None   \n",
      "1                                        \n",
      "2   havana_gene \"OTTHUMG00000189354.1\"   \n",
      "3   havana_gene \"OTTHUMG00000189354.1\"   \n",
      "4   havana_gene \"OTTHUMG00000189354.1\"   \n",
      "\n",
      "                                          16    17    18    19    20    21  \n",
      "0                                       None  None  None  None  None  None  \n",
      "1                                       None  None  None  None  None  None  \n",
      "2   havana_transcript \"OTTHUMT00000479422.1\"        None  None  None  None  \n",
      "3   havana_transcript \"OTTHUMT00000479422.1\"        None  None  None  None  \n",
      "4   havana_transcript \"OTTHUMT00000479422.1\"        None  None  None  None  \n",
      "\n",
      "[5 rows x 28 columns]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "from os import path\n",
    "\n",
    "curr_dir_path = str(pathlib.Path().absolute())\n",
    "data_path = curr_dir_path + \"/Data/\"\n",
    "\n",
    "def start_end_positions(file=\"chr21_annotations.csv\"):\n",
    "    '''\n",
    "    Function to read annotation given file for chromosome and get [start, end] base pair position of genes\n",
    "    :param file:\n",
    "            chromosome annotation file\n",
    "    :return: ndarray (num_genes, 2)\n",
    "            Array containing [start, end] base pair position on chromosome for each gene.\n",
    "    '''\n",
    "    final_chr = pd.read_csv(data_path + file, sep='\\t')\n",
    "    chr_gene = final_chr[final_chr['type'] == 'gene']\n",
    "    start_end_pos = pd.DataFrame(\n",
    "        {'start_pos': chr_gene['start'], 'end_pos': chr_gene['end']})  # shape: (875,2) for chr21\n",
    "    start_end_pos.reset_index(drop=True, inplace=True)\n",
    "    return start_end_pos\n",
    "\n",
    "\n",
    "def preprocess_chromosomeseq_file(chr_file=\"chr21.fa\", write_to_file=False):\n",
    "    '''\n",
    "    Function to remove special characters in chromosome sequence\n",
    "    :param chr_file: File containing nucleotide sequence for chromosome\n",
    "    :param write_to_file: bool, optional\n",
    "            Variable to control writing extracted information to txt file\n",
    "    :return: str\n",
    "            String containing nucleotide sequence for chromosome\n",
    "    '''\n",
    "    file_object = open(data_path + chr_file, \"r\")\n",
    "\n",
    "    chrm = file_object.read()\n",
    "    # print('Length of unprocessed chromosome: '+len(chrm))  # 47644190\n",
    "\n",
    "    chrm = chrm.replace('\\n', '')\n",
    "    #Fasta file contains name of chromosome (eg. '>chr21') at the start: removing it\n",
    "    if (chrm[5].isdigit()):\n",
    "        chrm = chrm[6:]\n",
    "    else:\n",
    "        chrm = chrm[5:]\n",
    "    # print('Length of processed chromosome: ', len(chrm))  # 46709983\n",
    "\n",
    "    #Append a blank at the beginning and end of string (since annotations are 1-indexed\n",
    "    chrm = ' '+chrm+' '   # len = 46709985\n",
    "    print('Length of processed chromosome: ', len(chrm))\n",
    "    if write_to_file or not path.exists(data_path+chr_file.replace(\"fa\", \"txt\")):\n",
    "        text_file = open(data_path + chr_file.replace(\"fa\", \"txt\"), \"w\")\n",
    "        text_file.write(chrm)\n",
    "        text_file.close()\n",
    "    return chrm\n",
    "\n",
    "\n",
    "def create_chromosome_annotations(chrm='chr21', write_to_file=False):\n",
    "    '''\n",
    "    Function to read annotation file for human genome sequence GRCh38 and get annotations for specified chromosome\n",
    "    :param chrm: str\n",
    "            chromosome information to be extracted\n",
    "    :param write_to_file: bool, optional\n",
    "            Variable to control writing extracted information to file\n",
    "    :return: None\n",
    "    '''\n",
    "    ann_file = \"gencode.v34.annotation.gtf\"\n",
    "    col_names = ['chr_name', 'source', 'type', 'start', 'end', '.', 'strand', ',', 'other']\n",
    "    ann = pd.read_csv(data_path + ann_file, sep='\\t', header=None)\n",
    "    ann.columns = col_names\n",
    "\n",
    "    # print(ann.shape)  # (2912496, 9)\n",
    "    # print(ann.head())\n",
    "\n",
    "    chr_no = ann[ann['chr_name'] == chrm]\n",
    "    del chr_no['.']\n",
    "    del chr_no[',']\n",
    "    #print(chr_no.head())\n",
    "\n",
    "    other = list(map(lambda x: x.split(';'), chr_no['other']))\n",
    "    df_other = pd.DataFrame(other)\n",
    "    df_other[0] = list(map(lambda x: x[8:].strip('\"'), df_other[0]))\n",
    "    df_other[1] = list(map(lambda x: x[14:].strip('\"') , df_other[1]))\n",
    "    del chr_no['other']\n",
    "\n",
    "    chr_no.reset_index(drop=True, inplace=True)\n",
    "    df_other.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    final_chr = pd.concat([chr_no, df_other], axis=1)\n",
    "    print(final_chr.head())\n",
    "    if write_to_file or not path.exists(data_path + chrm + '_annotations.csv'):\n",
    "        final_chr.to_csv(data_path + chrm + '_annotations.csv', index=False, sep='\\t')\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    #chrm = preprocess_chromosomeseq_file(\"chr21.fa\", write_to_file=False)  #read fasta file for chromosome, which generate the txt files\n",
    "    create_chromosome_annotations(chrm='chr21', write_to_file=True) #get annotations for particular chromosome\n",
    "    #file_object = open(data_path + \"chr21.txt\", \"r\")\n",
    "    #chrm = file_object.read()\n",
    "    #start_end_pos = start_end_positions(\"chr21_annotations.csv\")\n",
    "    #print(start_end_pos.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genes_exons_df(chrm, start_end_pos, write_to_file=False):\n",
    "    '''\n",
    "    Function to read positions and chromosome file and return the gene sequences\n",
    "    :param chrm: str\n",
    "            Nucleotide sequence of chromosome\n",
    "    :param start_end_pos: data frame\n",
    "            Start and end positions of each gene in the chromosome\n",
    "    :param write_to_file: bool, optional\n",
    "            Variable to control writing extracted information to file\n",
    "    :return: data frame (num_genes, 3)\n",
    "\n",
    "    '''\n",
    "    indices = list(range(0, len(start_end_pos)))\n",
    "    column_names = ['gene_sequence', 'gene_ranges', 'exon_ranges']\n",
    "    df = pd.DataFrame(index=indices, columns=column_names)\n",
    "\n",
    "    # get -300 and +300 nucleotides around each gene\n",
    "    df['gene_sequence'] = list(map(lambda x: chrm[start_end_pos['start_pos'][x]-300:start_end_pos['end_pos'][x]+300]\n",
    "                                   , indices) )\n",
    "    df['gene_ranges'] = list(zip(start_end_pos['start_pos'], start_end_pos['end_pos']))\n",
    "\n",
    "    gene_lists = list(map(lambda x: df.iloc[x]['gene_sequence'], indices))\n",
    "    df['exon_ranges'] = list(map(lambda x: get_exon_boundaries(x), gene_lists))\n",
    "\n",
    "    exon_lengths = [[x[1] - x[0] for x in ranges] for ranges in df['exon_ranges'] if (len(ranges) > 0)]\n",
    "    max_exon_lengths = list(map(max, exon_lengths))\n",
    "    min_exon_lengths = list(map(min, exon_lengths))\n",
    "    no_of_exons = list(map(len, exon_lengths))\n",
    "\n",
    "\n",
    "\n",
    "    meta_info = pd.DataFrame(list(zip(no_of_exons, max_exon_lengths, min_exon_lengths, exon_lengths)),\n",
    "                      columns=['no_of_exons', 'max_exon_length', 'min_exon_length', 'exon_lengths'])\n",
    "    meta_info.to_csv(data_path + 'chr21_exons_info.csv', index=False, sep='\\t')\n",
    "    if write_to_file:  # if file doesn't exist\n",
    "        df.to_csv(data_path + 'chr21_genes_exons.csv', index=False, sep='\\t')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def start_end_positions(file=\"chr21_annotations.csv\"):\n",
    "    \"\"\"\n",
    "    Function to read annotation given file for chromosome and get [start, end] base pair position of genes\n",
    "    :param file:\n",
    "            chromosome annotation file\n",
    "    :return: ndarray (num_genes, 2)\n",
    "            Array containing [start, end] base pair position on chromosome for each gene.\n",
    "    \"\"\"\n",
    "    final_chr = pd.read_csv(data_path + file, sep='\\t')\n",
    "    chr_gene = final_chr[final_chr['type'] == 'gene']\n",
    "    start_end_pos = pd.DataFrame(\n",
    "        {'start_pos': chr_gene['start'], 'end_pos': chr_gene['end']})  # shape: (875,2) for chr21\n",
    "    start_end_pos.reset_index(drop=True, inplace=True)\n",
    "    return start_end_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_chromosomeseq_file(chr_file=\"chr21.fa\", write_to_file=False):\n",
    "    '''\n",
    "    Function to remove special characters in chromosome sequence\n",
    "    :param chr_file: File containing nucleotide sequence for chromosome\n",
    "    :param write_to_file: bool, optional\n",
    "            Variable to control writing extracted information to file\n",
    "    :return: str\n",
    "            String containing nucleotide sequence for chromosome\n",
    "    '''\n",
    "    file_object = open(data_path + chr_file, \"r\")\n",
    "\n",
    "    chrm = file_object.read()\n",
    "    chrm = chrm.replace('\\n', '')\n",
    "    # print(len(chr21))  # 47644190\n",
    "    if write_to_file:\n",
    "        text_file = open(data_path + chr_file.replace(\"fa\", \"txt\"), \"w\")\n",
    "        text_file.write(chrm)\n",
    "        text_file.close()\n",
    "    return chrm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/asmitapoddar/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (25,26) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start,end shape: (875, 2)\n"
     ]
    }
   ],
   "source": [
    "import pathlib\n",
    "import pandas as pd\n",
    "from os import path\n",
    "import json\n",
    "import itertools\n",
    "\n",
    "curr_dir_path = str(pathlib.Path().absolute())\n",
    "data_path = curr_dir_path + \"/Data/\"\n",
    "\n",
    "def gene_start_end_positions(chrm_ann):\n",
    "    '''\n",
    "    Function to read annotation for given chromosome and get [start, end] base pair position of genes\n",
    "    :param chrm_ann: data frame\n",
    "            Annotations for given chromosome\n",
    "    :return: ndarray (num_genes, 2)\n",
    "            Array containing [start, end] base pair position on chromosome for each gene.\n",
    "    '''\n",
    "    chr_gene = chrm_ann[chrm_ann['type'] == 'gene']\n",
    "    gene_start_end_pos = pd.DataFrame(\n",
    "        {'start_pos': chr_gene['start'], 'end_pos': chr_gene['end']})  # shape: (875,2) for chr21\n",
    "    gene_start_end_pos.reset_index(drop=True, inplace=True)\n",
    "    print('Start,end shape:', gene_start_end_pos.shape)\n",
    "    return gene_start_end_pos\n",
    "\n",
    "def get_indices_of_table(df, last_index):\n",
    "    '''\n",
    "    Function to\n",
    "    :param df: data frame\n",
    "    :return: list of indexes of the data frame\n",
    "    '''\n",
    "    indices = pd.Index.to_list(df.index)\n",
    "    indices.append(last_index)\n",
    "\n",
    "    return indices\n",
    "\n",
    "def no_transcripts_per_gene(cur, nex, chrm_ann):\n",
    "    '''\n",
    "    Function to\n",
    "    :param cur:\n",
    "    :param nex:\n",
    "    :param chrm_ann:\n",
    "    :return:\n",
    "    '''\n",
    "    gene_table = chrm_ann.iloc[cur:nex]\n",
    "    transcript_counts_per_gene=gene_table['type'].value_counts()['transcript']\n",
    "\n",
    "    return transcript_counts_per_gene\n",
    "\n",
    "def get_chunk(cur, nex, chrm_ann, type):\n",
    "    '''\n",
    "    Function to get subset of the annotation data frame according to condition\n",
    "    :param cur: int\n",
    "            Start index\n",
    "    :param nex: int\n",
    "            End index\n",
    "    :param chrm_ann: data frame\n",
    "            Annotation data frame\n",
    "    :return: data frame\n",
    "            Subset of data frame\n",
    "    '''\n",
    "    table = chrm_ann.iloc[cur:nex]\n",
    "    df = table[table['type'] == type]\n",
    "    return df\n",
    "\n",
    "def create_dict(chrm_seq, chrm_ann):\n",
    "    '''\n",
    "    Function to create the json file\n",
    "    :param chrm_seq: string\n",
    "            Nucleotide sequence of the chromosome\n",
    "    :param chrm_ann: data frame\n",
    "            Annotation file for the chromosome\n",
    "    :return: dictionary\n",
    "        Dictionary to be stored as the json file\n",
    "    '''\n",
    "    chrm_gene = chrm_ann[chrm_ann['type'] == 'gene']\n",
    "\n",
    "    gene_ids = list(chrm_gene['0'])  #[l.strip('\"') for l in\n",
    "\n",
    "    gene_start_end_pos = gene_start_end_positions(chrm_ann)\n",
    "    gene_bounds = list(zip(gene_start_end_pos.start_pos, gene_start_end_pos.end_pos))\n",
    "\n",
    "    gene_strand = list(chrm_gene['strand'])\n",
    "    indices = list(range(0, len(gene_start_end_pos)))\n",
    "    gene_sequence = list(map(lambda x: chrm_seq[gene_start_end_pos['start_pos'][x]:gene_start_end_pos['end_pos'][x]]\n",
    "            , indices))\n",
    "    gene_indices = get_indices_of_table(chrm_gene, len(chrm_ann))\n",
    "\n",
    "    my_dictionary = {'main': []}\n",
    "\n",
    "    for i in range(0,len(chrm_gene)):\n",
    "        cur = gene_indices[i]\n",
    "        nex = gene_indices[i+1]\n",
    "        lis = {}\n",
    "\n",
    "        lis.update({'gene_id': gene_ids[i]})\n",
    "        lis.update({'gene_strand': gene_strand[i]})\n",
    "        lis.update({'gene_bounds': gene_bounds[i]})\n",
    "        lis.update({'gene_sequence':gene_sequence[i]})  #take into account reverse complementarity\n",
    "\n",
    "        no_transcripts = no_transcripts_per_gene(cur, nex, chrm_ann)\n",
    "        lis.update({'no_of_transcripts': int(no_transcripts)})\n",
    "\n",
    "        transcripts = []\n",
    "        transcript = get_chunk(cur, nex, chrm_ann, 'transcript')\n",
    "        transcript_ids = [l.strip('\"') for l in list(transcript['1'])]\n",
    "        transcript_ranges = list(zip(transcript['start'], transcript['end']))\n",
    "        transcript_indices = get_indices_of_table(transcript, nex)\n",
    "\n",
    "        for j in range(0,len(transcript_ranges)):\n",
    "\n",
    "            list_transcript = {}\n",
    "\n",
    "            list_transcript.update({'transcript_id': transcript_ids[j]})\n",
    "            list_transcript.update({'transcript_range': transcript_ranges[j]})\n",
    "\n",
    "            exon = get_chunk(transcript_indices[j], transcript_indices[j+1], chrm_ann, 'exon')\n",
    "            exon_ranges = list(zip(exon['start'], exon['end']))\n",
    "\n",
    "            list_transcript.update({'no_of_exons': int(len(exon_ranges))})\n",
    "            exons = []\n",
    "            list_exon_range = {}\n",
    "\n",
    "            for er in exon_ranges:\n",
    "                list_exon_range.update({'exon_ranges': er})\n",
    "                exons.append({'exon_ranges': er})\n",
    "\n",
    "            list_transcript.update({'exons': exons})\n",
    "            transcripts.append(list_transcript)\n",
    "\n",
    "        lis.update({'transcripts': transcripts})\n",
    "        my_dictionary['main'].append(lis)\n",
    "\n",
    "    return my_dictionary\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    chrm_txt_file = \"chr21.txt\"\n",
    "    chrm_ann_file = \"chr21_annotations.csv\"\n",
    "\n",
    "    file_object = open(data_path + chrm_txt_file, \"r\")\n",
    "    chrm_seq = file_object.read()\n",
    "    chrm_ann = pd.read_csv(data_path + chrm_ann_file, sep='\\t')\n",
    "\n",
    "    #with open(data_path+\"data.json\") as f:\n",
    "    #        data = json.load(f)\n",
    "    my_dictionary = create_dict(chrm_seq, chrm_ann)\n",
    "    with open(data_path+chrm_txt_file.replace('.txt','')+'_data.json', 'w') as file:\n",
    "        json.dump(my_dictionary, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_start_end_positions(chrm_ann):\n",
    "    '''\n",
    "    Function to read annotation for given chromosome and get [start, end] base pair position of genes\n",
    "    :param chrm_ann: data frame\n",
    "            Annotations for given chromosome\n",
    "    :return: ndarray (num_genes, 2)\n",
    "            Array containing [start, end] base pair position on chromosome for each gene.\n",
    "    '''\n",
    "    chr_gene = chrm_ann[chrm_ann['type'] == 'gene']\n",
    "    gene_start_end_pos = pd.DataFrame(\n",
    "        {'start_pos': chr_gene['start'], 'end_pos': chr_gene['end']})  # shape: (875,2) for chr21\n",
    "    gene_start_end_pos.reset_index(drop=True, inplace=True)\n",
    "    print('Start,end shape:', gene_start_end_pos.shape)\n",
    "    return gene_start_end_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/asmitapoddar/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (24,25) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "chrm_txt_file = \"chr21.txt\"\n",
    "chrm_ann_file = \"chr21_annotations.csv\"\n",
    "\n",
    "file_object = open(data_path + chrm_txt_file, \"r\")\n",
    "chrm_seq = file_object.read()\n",
    "chrm_ann = pd.read_csv(data_path + chrm_ann_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start,end shape: (875, 2)\n"
     ]
    }
   ],
   "source": [
    "df = gene_start_end_positions(chrm_ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>5011799</td>\n",
       "      <td>5017145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>5022493</td>\n",
       "      <td>5040666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>5073458</td>\n",
       "      <td>5087867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>5079294</td>\n",
       "      <td>5128413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5116343</td>\n",
       "      <td>5133805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start_pos  end_pos\n",
       "0    5011799  5017145\n",
       "1    5022493  5040666\n",
       "2    5073458  5087867\n",
       "3    5079294  5128413\n",
       "4    5116343  5133805"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr_name',\n",
       " 'source',\n",
       " 'type',\n",
       " 'start',\n",
       " 'end',\n",
       " '0',\n",
       " '1',\n",
       " '2',\n",
       " '3',\n",
       " '4',\n",
       " '5',\n",
       " '6',\n",
       " '7',\n",
       " '8',\n",
       " '9',\n",
       " '10',\n",
       " '11',\n",
       " '12',\n",
       " '13',\n",
       " '14',\n",
       " '15',\n",
       " '16',\n",
       " '17',\n",
       " '18',\n",
       " '19',\n",
       " '20',\n",
       " '21']"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrm_gene = chrm_ann[chrm_ann['type'] == 'gene']\n",
    "\n",
    "list(chrm_gene.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    " gene_ids = [l.strip('\"') for l in list(chrm_gene['0'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start,end shape: (875, 2)\n"
     ]
    }
   ],
   "source": [
    "gene_start_end_pos = gene_start_end_positions(chrm_ann) \n",
    "l = list(zip(gene_start_end_pos.start_pos, gene_start_end_pos.end_pos))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "indices = list(range(0, len(gene_start_end_pos)))\n",
    "gene_sequence = list(map(lambda x: chrm_seq[gene_start_end_pos['start_pos'][x]:gene_start_end_pos['end_pos'][x]]\n",
    "            , indices))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
